### DISCUSSION TITLE
Does the shaping result include the space around the glyph?

### DISCUSSION THREAD

**Discussion by kaiwk** [harfbuzz/harfbuzz#2791] [Q&A] [answered]
![hello](https://user-images.githubusercontent.com/10190397/103359331-ac687500-4af2-11eb-8bc1-7b51cd1fec04.png)

I use the `hb-view` to generate the picture above, and I find there are space above/below the glyph. Does that mean the ascender/descender also include the space?

In Chrome, when i set `padding: 0px;`, there are always some spaces between the text and border, is that the same reason? 

Thanks and happy New Year!

**Comment by khaledhosny** [comment]
This is controlled by font (and application) line vertical metrics. HarfBuzz can be used to query the font for its metrics, but otherwise has no influence over how the application uses them.

**Comment by alerque** [comment, marked as answer]
As Khaled noted, this is entirely up to the font & application. Harfbuzz can only report what the font declares, and for *many* fonts there may be extra space besides what is being declared.

Additionally browsers are notorious for adding even more spacing in various places (particularly vertical space before and after lines of text) that are not part of the font or shaping process. There are some efforts underway to bring this under control, but by default you'll find most browsers are terribly behaved (in what I think was a misguided attempt to make things look okay even when page designers weren't thinking about the consequences of setting things like `padding: 0`. See for example [this leading control proposal](https://github.com/w3c/csswg-drafts/issues/3240) that would tame some of issues with leading. Articles like [this one](https://css-tricks.com/how-to-tame-line-height-in-css/) have some tips for how to survive in the mean time.
